Documentation/Contributors/CodingGuide/README.md

```diff
@@ -0,0 +1,662 @@
+    if (!defined(c)) {
+        throw new DeveloperError('cartesian is required.');
+    }
+    //>>includeEnd('debug');
+
+    return Math.max(c.x, c.y, c.z); // Works in debug. Fails in release!
+};
+```
+* Throw Cesium's `RuntimeError` for an error that will not be known until runtime.  Unlike developer errors, runtime error checks are not optimized out of release builds.
+```javascript
+if (typeof WebGLRenderingContext === 'undefined') {
+    throw new RuntimeError('The browser does not support WebGL.');
+}
+```
+* :art: Exceptions are exceptional.  Avoid throwing exceptions, e.g., if a polyline is only provided one position, instead of two or more, don't render it instead of throwing an exception.
+
+### `result` Parameters and Scratch Variables
+
+:speedboat: In JavaScript, user-defined classes such as `Cartesian3` are reference types and are therefore allocated on the heap.  Frequently allocating these types causes a significant performance problem because it creates GC pressure, which causes the Garbage Collector to run longer and more frequently.
+
+Cesium uses required `result` parameters to avoid implicit memory allocation.  For example,
+```javascript
+var sum = Cartesian3.add(v0, v1);
+```
+would have to implicitly allocate a new `Cartesian3` object for the returned sum.  Instead, `Cartesian3.add` requires a `result` parameter:
+```javascript
+var result = new Cartesian3();
+var sum = Cartesian3.add(v0, v1, result); // result and sum reference the same object
+```
+This makes allocations explicit to the caller, which allows the caller to, for example, reuse the result object in a file-scoped scratch variable:
+```javascript
+var scratchDistance = new Cartesian3();
+
+Cartesian3.distance = function(left, right) {
+    Cartesian3.subtract(left, right, scratchDistance);
+    return Cartesian3.magnitude(scratchDistance);
+};
+```
+The code is not as clean, but the performance improvement is often dramatic.
+
+As described below, `from` constructors also use optional `result` parameters.
+
+## Classes
+
+* :art: Classes should be **cohesive**; a class should represent one abstraction.
+* :art: Classes should be **loosely coupled**; two classes should not be entangled and rely on each other's implementation details; they should communicate through well defined interfaces.
+
+### Constructor Functions
+
+* Create a class by creating a constructor function:
+```javascript
+var Cartesian3 = function(x, y, z) {
+    this.x = defaultValue(x, 0.0);
+    this.y = defaultValue(y, 0.0);
+    this.z = defaultValue(z, 0.0);
+};
+```
+* Create an instance of a class (an _object_) by calling the constructor function with `new`:
+```javascript
+var p = new Cartesian3(1.0, 2.0, 3.0);
+```
+* :speedboat: Assign to all the property members of a class in the constructor function to allow V8 to use a hidden class and avoid entering dictionary mode.  Assign `undefined` if no initial value makes sense.  Do not add properties to an object, e.g.,
```

Comment:
While Cesium runs best in Chrome, I'm not sure it's necessary to call out V8 for things like this.  This is true of all modern browsers.